spec/lang/step/operators.md

```diff
@@ -346,28 +346,67 @@ impl<M: Memory> Machine<M> {
-    fn prepare_for_rel_op(val: Value<M>) -> (Int, Int) {
+    /// Prepare a value for being compared with a `RelOp`. This involves turning it into an `Int`.
+    fn prepare_for_int_rel_op(val: Value<M>) -> Int {
         match val {
-            Value::Int(i) => (i, Int::ZERO),
-            Value::Bool(b) => (if b { Int::from(1) } else { Int::from(0) }, Int::ZERO),
-            Value::Ptr(Pointer { thin_pointer, metadata: None}) => (thin_pointer.addr, Int::ZERO),
-            Value::Ptr(Pointer { thin_pointer, metadata: Some(PointerMeta::ElementCount(i))}) => (thin_pointer.addr, i),
-            Value::Ptr(Pointer { thin_pointer, metadata: Some(PointerMeta::VTablePointer(v))}) => (thin_pointer.addr, v.addr),
-            _ => panic!("invalid value for relational operator"),
+            Value::Int(i) => i,
+            Value::Bool(b) => if b { Int::from(1) } else { Int::from(0) },
+            _ => panic!("invalid value for integer relational operator"),
         }
     }
+
+    fn eval_ptr_rel_op(rel: RelOp, left: Pointer<M::Provenance>, right: Pointer<M::Provenance>) -> (Value<M>, Type) {
+        use RelOp::*;
+        let cmp = Self::compare_ptr(left, right);
+        match rel {
+            Lt => (Value::Bool(cmp.is_lt()), Type::Bool),
+            Gt => (Value::Bool(cmp.is_gt()), Type::Bool),
+            Le => (Value::Bool(cmp.is_le()), Type::Bool),
+            Ge => (Value::Bool(cmp.is_ge()), Type::Bool),
+            Eq => (Value::Bool(cmp.is_eq()), Type::Bool),
+            Ne => (Value::Bool(cmp.is_ne()), Type::Bool),
+            Cmp => {
+                let val = match cmp {
+                    std::cmp::Ordering::Less => -1,
+                    std::cmp::Ordering::Equal => 0,
+                    std::cmp::Ordering::Greater => 1,
+                };
+                (Value::Int(Int::from(val)), Type::Int(IntType::I8))
+            }
+        }
+    }
+    fn compare_ptr(left: Pointer<M::Provenance>, right: Pointer<M::Provenance>) -> std::cmp::Ordering {
+        let thin_cmp = left.thin_pointer.addr.cmp(&right.thin_pointer.addr);
+        let meta_cmp = match (left.metadata, right.metadata) {
+            (None, None) => std::cmp::Ordering::Equal,
+            (Some(PointerMeta::ElementCount(l)), Some(PointerMeta::ElementCount(r))) => l.cmp(&r),
+            (Some(PointerMeta::VTablePointer(l)), Some(PointerMeta::VTablePointer(r))) => l.addr.cmp(&r.addr),
+            _ => panic!("unmatching metadata in wide pointer comparasion"),
+        };
+        // Lexicographically compare on first the thin pointer and then the metadata
+        thin_cmp.then(meta_cmp)
+    }
+
     fn eval_bin_op(
         &self,
         BinOp::Rel(rel_op): BinOp,
         (left, l_ty): (Value<M>, Type),
         (right, _r_ty): (Value<M>, Type)
     ) -> Result<(Value<M>, Type)> {
-        let left = Self::prepare_for_rel_op(left);
-        let right = Self::prepare_for_rel_op(right);
+        match l_ty {
+            Type::Int(_) | Type::Bool => {
```

Comment:
I don't think it's worth having these in a shared match arm; `prepare_for_int_rel_op` isn't carrying its weight any more IMO.